pinot-core/src/main/java/org/apache/pinot/core/operator/filter/AndFilterOperator.java

```diff
@@ -59,13 +60,14 @@ protected BlockDocIdSet getTrues() {
   protected BlockDocIdSet getFalses() {
     List<BlockDocIdSet> blockDocIdSets = new ArrayList<>(_filterOperators.size());
     for (BaseFilterOperator filterOperator : _filterOperators) {
-      if (filterOperator.isResultEmpty()) {
-        blockDocIdSets.add(new MatchAllDocIdSet(_numDocs));
+      if (_nullHandlingEnabled) {
+        blockDocIdSets.add(
+            new OrDocIdSet(Arrays.asList(filterOperator.getTrues(), filterOperator.getNulls()), _numDocs));
```

Comment:
After a second thought, will this be even slower than the original one? Here we are essentially creating an `OrDocIdSet` on every predicate.

We want to check whether the filter matches all or empty, then optimize the `DocIdSet` accordingly.

Reply:
I think that makes sense as an optimization, added the cases.

I don't the second point is a regression - previously .getFalses() created an `OrDocIdSet` for every predicate anyway, so we'd see something like this:
```
OrDocIdSet(NotDocIdSet(OrDocIdSet(...), NotDocIdSet(OrDocIdSet(...))
```

with the change in this PR it would instead be:
```
NotDocIdSet(AndDocIdSet(OrDocIdSet(...), OrDocIdSet(...)))
```

This specific query speedup is very apparent, using a quickstart dataset but w/ 60k rows I did a quick comparison:
```
q1: select count(*) from fineFoodReviews where NOT regexp_like("Text", 'happen to be allergic to it')
q2: select count(*) from fineFoodReviews where NOT text_match("Text", '"happen to be allergic to it"')
q3: select count(*) from fineFoodReviews where NOT (text_match("Text", '"happen to be allergic to it"') AND regexp_like("Text", 'happen to be allergic to it'))

q1: 39ms
q2: 5ms
q3: 6ms
```

Without the change, q3 latency is >= q1. I'm happy to set up a microbench if you think it'd be helpful.

Reply:
I see. Thanks for the explanation! Basically we are always taking the overhead in `getFalses()`. I feel we should also try to optimize the default `getFalses()` to check if `getNulls()` returns `EmptyDocIdSet`, and skip creating `OrDocIdSet` if it is empty. The same optimization can also be applied here